packages/polkadart_scale_codec/lib/src/core/codec.dart

```diff
@@ -453,18 +453,84 @@ class Codec {
 
   ///
   /// Decodes Bit Sequence
-  List<int> _decodeBitSequence(Source source) {
-    var length = (source.compactLength() / 8).ceil();
-    return source.bytes(length).toList();
+  List<int> _decodeBitSequence(BitSequenceType def, Source source) {
+    // get compacted length for filling result
+    final int length = source.compactLength();
+
+    // get the capacity of the bit sequence
+    final int capacity = max(length, _getBitSequenceCapacity(def.bitStoreType));
+
+    // get the encoded bytes from the source
+    final List<int> bytes = source.bytes((capacity / 8).ceil()).toList();
+
+    // get the extra bit length so to form a 4 * Uint8List
+    final int extraBitLength = capacity - bytes.length;
+
+    // add the 0 to the bits
+    for (var i = 0; i < extraBitLength; i++) {
+      bytes.add(0);
+    }
+
+    final BitArray bitArray = BitArray.fromUint8List(Uint8List.fromList(bytes));
+
+    final List<int> bits = List.filled(length, 0);
+
+    for (var i = 0; i < length; i++) {
+      bits[i] = bitArray[i] ? 1 : 0;
+    }
+
+    return bits;
   }
 
   ///
   /// Encodes Bit Sequence
-  void _encodeBitSequence(dynamic bits, ScaleCodecEncoder encoder) {
-    assertionCheck(
-        bits is List<int>, 'BitSequence can have bits of type List<int> only.');
-    encoder.compact(bits.length * 8);
-    encoder.bytes(bits);
+  void _encodeBitSequence(
+      BitSequenceType def, dynamic bits, ScaleCodecEncoder encoder) {
+    assertionCheck(bits is List<int>,
+        'BitSequence can have bits of type `List<int>` only.');
+
+    int capacity = max(bits.length, _getBitSequenceCapacity(def.bitStoreType));
+
+    capacity = (capacity / 8).ceil() * 8;
+
+    encoder.compact(bits.length);
+
+    final bitArray = BitArray(capacity);
+    for (var i = 0; i < bits.length; i++) {
+      switch (bits[i]) {
+        case 0:
+          bitArray[i] = false;
+          break;
+        case 1:
+          bitArray[i] = true;
+          break;
+        default:
+          throw UnexpectedCaseException(
+              'Invalid byte, unable to encode $bits.');
+      }
+    }
+
+    encoder.bytes(
+        bitArray.byteBuffer.asUint8List(0, (capacity / 8).ceil()).toList());
+  }
+
+  int _getBitSequenceCapacity(int bitStoreTypeIndex) {
+    // setting default capacity to 8
+    if (_types[bitStoreTypeIndex].kind != TypeKind.Primitive) {
+      return 8;
+    }
+    final type = _types[bitStoreTypeIndex] as PrimitiveType;
+    final name = type.primitive.name.toLowerCase();
+    if (name.startsWith('u') == false) {
+      throw UnexpectedCaseException(
+          'Expected unsigned bit with `u` prefix, but got $name which can\'t be decoded');
+    }
+    if (name.length == 1 || int.parse(name.substring(1)) % 8 != 0) {
+      throw UnexpectedCaseException(
+          'Expected unsigned bit with valid bitSize, but got ${name.length == 1 ? '' : name.substring(1)} which can\'t be decoded.');
+    }
```

Comment:
I think it's weird to have an empty string when we actually got `'u'`. `name` should provide context for the user on what's wrong
```suggestion
          'Expected unsigned bit with valid bitSize, but got $name which can\'t be decoded.');
    }
```